apps/manager_app/joystick_training.c

```diff
@@ -0,0 +1,255 @@
+      MANAGER_TRAIN_JOYSTICK_REQUEST_INITIATE_TAG != req->which_request) {
+    send_training_error(1);    // TODO: use correct data-error code
+    return false;
+  }
+
+  // TODO: check the request with what device knows
+  return true;
+}
+
+static void send_training_error(uint32_t error_code) {
+  manager_train_joystick_response_t training =
+      MANAGER_TRAIN_JOYSTICK_RESPONSE_INIT_ZERO;
+  training.which_response = MANAGER_TRAIN_JOYSTICK_RESPONSE_COMMON_ERROR_TAG;
+  training.common_error.which_error =
+      ERROR_COMMON_ERROR_DEVICE_SETUP_REQUIRED_TAG;
+  training.common_error.device_setup_required = error_code;
+  send_msg_to_host(&training);
+}
+
+static void send_msg_to_host(manager_train_joystick_response_t *training) {
+  uint8_t payload[MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE] = {0};
+  manager_result_t msg = MANAGER_RESULT_INIT_ZERO;
+  size_t msg_size = 0;
+  msg.which_response = MANAGER_RESULT_TRAIN_JOYSTICK_TAG;
+
+  memcpy(
+      &msg.train_joystick, training, sizeof(manager_train_joystick_response_t));
+  encode_manager_result(
+      &msg, payload, MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE, &msg_size);
+  usb_send_msg(payload, msg_size);
+}
+
+static bool training_step(const joystick_step_t *step) {
+  evt_status_t events;
+  joystick_train_init(step->instruction, step->user_action);
+  do {
+    events = get_events(EVENT_CONFIG_UI, MAX_INACTIVITY_TIMEOUT);
+    if (events.p0_event.abort_evt) {
+      usb_clear_event();
+      return false;
+    }
+  } while (!events.ui_event.event_occured);
+  core_status_set_flow_status(step->status);
+  return true;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void manager_joystick_training(manager_query_t *query) {
+  const joystick_step_t steps[JOYSTICK_TRAIN_STEPS] = {
+      {.instruction = ui_text_joystick_up,
+       .user_action = JS_ACTION_UP,
+       .status = MANAGER_TRAIN_JOYSTICK_UP},
+      {.instruction = ui_text_joystick_right,
+       .user_action = JS_ACTION_RIGHT,
+       .status = MANAGER_TRAIN_JOYSTICK_RIGHT},
+      {.instruction = ui_text_joystick_down,
+       .user_action = JS_ACTION_DOWN,
+       .status = MANAGER_TRAIN_JOYSTICK_DOWN},
+      {.instruction = ui_text_joystick_left,
+       .user_action = JS_ACTION_LEFT,
+       .status = MANAGER_TRAIN_JOYSTICK_LEFT},
+      {.instruction = ui_text_joystick_center,
+       .user_action = JS_ACTION_CENTER,
+       .status = MANAGER_TRAIN_JOYSTICK_CENTER}};
+
+  core_status_set_device_waiting_on(CORE_DEVICE_WAITING_ON_IDLE);
+
+  if (false == validate_training_request(query)) {
+    return;
+  }
+
+  uint8_t step_index = 0;
+  for (step_index = 0; step_index < JOYSTICK_TRAIN_STEPS; step_index++) {
+    const joystick_step_t *step = &steps[step_index];
+    if (false == training_step(step)) {
+      return;
```

Comment:
Can you write a brief comment related to early return.

Reply:
added